src/core/packages/elasticsearch/client-server-mocks/src/mocks.ts

```diff
@@ -122,67 +121,176 @@ const createMockedApi = <
-  client.child = jest.fn().mockImplementation(() => createInternalClientMock());
+function buildLazyMockFromShape(shape: ShapeNode, res?: Promise<unknown>): any {
+  if (shape.type !== 'object') return {};
+  const target: Record<string, any> = {};
 
-  const mockGetter = (obj: Record<string, any>, propertyName: string) => {
-    Object.defineProperty(obj, propertyName, {
+  const defineLazyMethod = (obj: Record<string, any>, key: string) => {
+    Object.defineProperty(obj, key, {
       configurable: true,
-      enumerable: false,
-      get: () => jest.fn(),
-      set: undefined,
+      enumerable: true,
+      get() {
+        const fn = createMockedApi();
+        fn.mockImplementation(() => res ?? createSuccessTransportRequestPromise({}));
+        Object.defineProperty(obj, key, {
+          value: fn,
+          configurable: true,
+          enumerable: true,
+          writable: true,
+        });
+        return fn;
+      },
+      set(value) {
+        Object.defineProperty(obj, key, {
+          value,
+          configurable: true,
+          enumerable: true,
+          writable: true,
+        });
+      },
     });
   };
 
-  // `on`, `off`, and `once` are properties without a setter.
-  // We can't `client.diagnostic.on = jest.fn()` because the following error will be thrown:
-  // TypeError: Cannot set property on of #<Client> which has only a getter
-  mockGetter(client.diagnostic, 'on');
-  mockGetter(client.diagnostic, 'off');
-  mockGetter(client.diagnostic, 'once');
-  client.transport = {
-    request: jest.fn(),
+  const defineLazyObject = (obj: Record<string, any>, key: string, childShape: ShapeNode) => {
+    Object.defineProperty(obj, key, {
+      configurable: true,
+      enumerable: true,
+      get() {
+        const child = buildLazyMockFromShape(childShape, res);
+        Object.defineProperty(obj, key, {
+          value: child,
+          configurable: true,
+          enumerable: true,
+          writable: true,
+        });
+        return child;
+      },
+    });
   };
 
-  return client as DeeplyMockedApi<Client>;
+  for (const [key, node] of Object.entries(shape.props)) {
+    if (node.type === 'method') {
+      defineLazyMethod(target, key);
+    } else if (node.type === 'object') {
+      defineLazyObject(target, key, node);
+    }
+  }
+
+  // Special cases based on prior behavior
+  Object.defineProperty(target, 'diagnostic', {
+    configurable: true,
+    enumerable: false,
+    get() {
+      const d: any = {};
+      for (const k of ['on', 'off', 'once']) {
+        Object.defineProperty(d, k, {
+          configurable: true,
+          enumerable: false,
+          get: () => jest.fn(),
+        });
+      }
+      Object.defineProperty(target, 'diagnostic', {
+        value: d,
+        configurable: true,
+        enumerable: false,
+        writable: true,
+      });
+      return d;
+    },
+  });
+
+  Object.defineProperty(target, 'transport', {
```

Comment:
Just out of curiosity, I don't fully grasp the idea behind this custome property behavior. Could you explain?
Besides, I see that `diagnostic` and `helpers` have been removed from `omittedProps`, but here you're customising `diagnostic` and `transport`.

Reply:
This part is similar to what we had before:
```
  // `on`, `off`, and `once` are properties without a setter.
  // We can't `client.diagnostic.on = jest.fn()` because the following error will be thrown:
  // TypeError: Cannot set property on of #<Client> which has only a getter
  mockGetter(client.diagnostic, 'on');
  mockGetter(client.diagnostic, 'off');
  mockGetter(client.diagnostic, 'once');
  client.transport = {
    request: jest.fn(),
  };
```
So, in both cases, we only keep the fields that we want to mock and for the rest, we just omit it in this custom implementation.

My understanding is that, we don't need to omit it anymore since we don't parse these fields unless we access them, and during the first access, we customize it similar to what we had before.